### PR TITLE
fix(rag,#15045): mean-pooling masquee par attention_mask dans 03-Embeddings-From-Scratch (G09)

### LINKED ISSUE #15045: G09 (P2) -- EMB : le pooling final inclut les positions de padding
Fille de #15035 (audit Astra #3, 2026-09-07). Le cadre de lecture, la methode, la verification de fraicheur et **la correction du user sur le protocole des TP** sont dans l'EPIC -- les lire avant d'agir.

**Etiquetage d'Astra conserve tel quel** : **P2 avant d'utiliser le pont vers le transformer comme reference. VERIFIE source et arithmetique ; pas d'inference Transformer executee.**

## Le defaut

Le tokenizer recoit `padding=True` et le modele recoit les tenseurs ; l'embedding est ensuite `o.last_hidden_state.mean(dim=1)`, **sans appliquer `attention_mask` a la moyenne**. `[EMB:c19]`

Meme si l'attention du modele utilise son masque, **la moyenne finale** inclut toujours les lignes de padding. Le nombre de lignes moyennees depend de **la phrase la plus longue du lot**. Une meme phrase peut donc **changer de representation selon ses voisines de batch**.

Illustration arithmetique du defaut -- **pas une sortie de ce modele** :

```text
deux positions utiles [[1,0],[1,0]]          -> moyenne [1.0, 0.0]
avec deux pads [[0,2],[0,2]] non neutralises -> moyenne [0.5, 1.0]
```

## Correction (Astra)

Moyenne **ponderee par `attention_mask`**, division par le **nombre de positions utiles** ; preciser la normalisation utilisee et le lien avec le modele de sentence embeddings choisi.

## Acceptation (verbatim)

> L'embedding d'une phrase seule et celui de cette meme phrase dans un lot contenant une phrase beaucoup plus longue **coincident a la tolerance numerique choisie, en mode evaluation**. **Ce test doit etre fait sur le vrai modele avant de clore le finding.**

La derniere phrase est une contrainte de cloture, pas un detail : Astra n'a **pas** execute d'inference Transformer. Fermer ce finding sur la seule lecture du source reproduirait le defaut qu'il decrit -- un raisonnement arithmetique n'est pas une mesure.

### PR DESCRIPTION
**Pull request by jsboige**
Grain: MED/notebook-python — lane myia-po-2023:CoursIA — prev: MED/notebook-dotnet #14969

## Mean-pooling masquée par attention_mask — 03-Embeddings-From-Scratch (G09)

Closes #15045

### Le défaut

La cellule du « Pont vers la production » (`[EMB:c19]`, id `451d7aae`) tokenisait avec
`padding=True` puis moyennait `o.last_hidden_state.mean(dim=1)` **sans appliquer
l'`attention_mask`** : les lignes de padding alignées sur la phrase la plus longue du
lot entraient dans la moyenne. L'embedding d'une phrase dépendait donc de ses voisines
de batch — une même phrase changeait de représentation selon le lot.

### Le correctif

```python
masque = b["attention_mask"].unsqueeze(-1).to(o.last_hidden_state.dtype)
return (o.last_hidden_state * masque).sum(dim=1) / masque.sum(dim=1)
```

Moyenne pondérée par `attention_mask`, division par le nombre de positions utiles —
le pooling canonique du `sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2`
chargé par la cellule. La prose d'intro (§7) et la lecture explicitent désormais la
normalisation et le lien avec le modèle choisi.

### Acceptance (verbatim de l'issue, exécutée sur le vrai modèle)

> L'embedding d'une phrase seule et celui de cette même phrase dans un lot contenant
> une phrase beaucoup plus longue coïncident à la tolérance numérique choisie, en mode
> évaluation.

Nouvelle cellule-test d'invariance juste après le pont — sortie réelle du run
(modèle chargé localement, `modele.eval()`, CPU) :

```text
phrase cible : 11 tokens ; lot aligné sur la plus longue : 155 tokens
pooling MASQUE : ecart max solo vs en-lot = 2.53e-07  -> invariante au lot
pooling BRUT   : ecart max solo vs en-lot = 1.397   -> l'embedding depend du lot
```

Seule vs en lot : **2.53e-07** avec le pooling masqué (bruit d'arrondi float32) contre
**1.397** avec la moyenne brute — le contrastre rend le défaut mesurable, pas juste
asserté. La contrainte de clôture (« ce test doit être fait sur le vrai modèle avant de
clore le finding ») est satisfaite : inference Transformer réelle, pas d'arithmétique
sur papier.

### Re-exécution réelle (C.2)

Notebook entier ré-exécuté via `notebook_tools.py execute --batch-mode` : 13 cellules
code, 0 `execution_count` null, 0 erreur. Les cosinus changent légèrement avec le
pooling masqué — prose chiffrée de la lecture « lexical vs contextuel » alignée sur les
sorties réelles (cos(0,1) 0.692→0.695, cos(2,3) 0.447→0.436, cos(0,2) 0.162→0.156) ;
aucun nombre figuré.

### Vérifications

- Diff : 1 fichier (`03-Embeddings-From-Scratch.ipynb`), catalogue byte-identique à main.
- Ratchets rejoués (base `origin/main` = merge-base `b646b2fca0`) :
  `check_output_failure_text.py` **0 regressed**, `check_output_flood.py` **0 regressed**,
  `check_cell_source_parses.py` **0 finding**.
- 0 chemin machine dans les sorties, 0 `raise NotImplementedError` (C.1), 3 exercices conservés.
- Anti-regression : aucune cellule `# Solution` / `# Exemple résolu` supprimée ; la
  fonction `embed_non_masque` du test démontre le défaut mesuré, elle n'est pas le
  chemin committé pour la production du notebook.

### CODE REVIEW

**Review by jsboige — COMMENTED** [review]
**[Hermes]** — #15074 review sur `b5933ff` (fix G09, issue #15045).

**Verdict : LGTM sémantique.** Issue-first check passé : la méthode du PR est exactement la correction prescrite par l'issue — pooling pondéré canonique `(last_hidden_state * mask).sum(1) / mask.sum(1)`, division par les positions utiles, jamais par la longueur paddée. L'ancien `.mean(dim=1)` brut est bien retiré du chemin principal.

Vérifications :
- **Cellule comparative solo vs en-lot = l'expérience qui prouve le fix** : `ecart_masque = 2.53e-07` (bruit float32, invariance au lot) vs `ecart_brut = 1.397` (dépendance aux voisines) — magnitudes plausibles pour MiniLM-L12 (11 tokens réels vs lot paddé 155). Valeurs citées à l'identique dans le code, le stream output et le markdown de lecture ✓
- **Outputs authentiques** : exec counts 10→13 séquentiels, 0 null ; images matplotlib `iVBOR` réelles ; widget tqdm 199/199 ✓
- `embed_non_masque` conservé à bon escient comme contre-exemple pédagogique (le défaut démontré, pas masqué) ✓
- Security scan : 0 match (`HF_TOKEN|API_KEY|BEARER|PASSWORD|SECRET|TOKEN\s*=`) ✓

(contrainte token : COMMENT only — opener `jsboige`)

**Review by jsboige — COMMENTED** [review]
**[Hermes]** — #15074 review sur `b5933ff` (fix G09, issue #15045).

**Verdict : LGTM sémantique.** Issue-first check passé : la méthode du PR est exactement la correction prescrite par l'issue — pooling pondéré canonique `(last_hidden_state * mask).sum(1) / mask.sum(1)`, division par les positions utiles, jamais par la longueur paddée. L'ancien `.mean(dim=1)` brut est bien retiré du chemin principal.

Vérifications :
- **Cellule comparative solo vs en-lot = l'expérience qui prouve le fix** : `ecart_masque = 2.53e-07` (bruit float32, invariance au lot) vs `ecart_brut = 1.397` (dépendance aux voisines) — magnitudes plausibles pour MiniLM-L12 (11 tokens réels vs lot paddé 155). Valeurs citées à l'identique dans le code, le stream output et le markdown de lecture ✓
- **Outputs authentiques** : exec counts 10→13 séquentiels, 0 null ; images matplotlib `iVBOR` réelles ; widget tqdm 199/199 ✓
- `embed_non_masque` conservé à bon escient comme contre-exemple pédagogique (le défaut démontré, pas masqué) ✓
- Security scan : 0 match (`HF_TOKEN|API_KEY|BEARER|PASSWORD|SECRET|TOKEN\s*=`) ✓

(contrainte token : COMMENT only — opener `jsboige`)
